src/java.base/share/classes/jdk/internal/foreign/LayoutRecordMapper.java

```diff
@@ -0,0 +1,641 @@
+import java.lang.foreign.PaddingLayout;
+import java.lang.foreign.SequenceLayout;
+import java.lang.foreign.ValueLayout;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.reflect.Array;
+import java.lang.reflect.RecordComponent;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.BinaryOperator;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+import static java.util.stream.Collectors.toMap;
+
+/**
+ * A record mapper that is matching a GroupLayout to match the components of a record.
+ *
+ * @param <T> the Record type
+ */
+public final class LayoutRecordMapper<T extends Record>
+        implements Function<MemorySegment, T> {
+
+    enum Allow {EXACT, BOXING, BOXING_NARROWING_AND_WIDENING}
+
+    public static final Allow ALLOW = Allow.EXACT;
+
+    private static final MethodHandles.Lookup LOOKUP = MethodHandles.lookup();
+    private static final MethodHandles.Lookup PUBLIC_LOOKUP = MethodHandles.publicLookup();
+
+    private final Class<T> type;
+    private final GroupLayout layout;
+    private final MethodHandle ctor;
+
+    public LayoutRecordMapper(Class<T> type,
+                              GroupLayout layout) {
+        this(type, layout, 0);
+    }
+
+    private LayoutRecordMapper(Class<T> type,
+                               GroupLayout layout,
+                               long offset) {
+        this.type = type;
+        this.layout = layout;
+
+        var recordComponents = Arrays.asList(type.getRecordComponents());
+        var recordComponentNames = recordComponents.stream()
+                .map(RecordComponent::getName)
+                .collect(Collectors.toSet());
+
+        var nameToLayoutMap = layout.memberLayouts().stream()
+                // Only consider named layouts
+                .filter(l -> l.name().isPresent())
+                // Only look at the relevant layouts which could ever be matched with a record component
+                .filter(l -> recordComponentNames.contains(l.name().orElseThrow()))
+                .collect(toMap(l -> l.name().orElseThrow(), Function.identity(), throwingMerger()));
+
+        var missingComponents = recordComponents.stream()
+                .map(RecordComponent::getName)
+                .filter(l -> !nameToLayoutMap.containsKey(l))
+                .toList();
+
+        if (!missingComponents.isEmpty()) {
+            throw new IllegalArgumentException("There is no mapping for " +
+                    missingComponents + " in " + type.getName() +
+                    "(" + String.join(", ", recordComponents.stream().map(RecordComponent::getName).collect(Collectors.joining(", "))) + ")" +
```

Comment:
Two joins, String.join and Collectors.joining() ??

Reply:
Also, I think this can be simplified - e.g. "there is no mapping for [list of component names] in [group layout]". E.g. just printing the group layout will be enough, I think.